core/database_api.php

```diff
@@ -332,9 +334,10 @@ function db_query_bound() {
  * @param array   $p_arr_parms Array of parameters matching $p_query.
  * @param integer $p_limit     Number of results to return.
  * @param integer $p_offset    Offset query results for paging.
+ * @param boolean $p_pop_param Set to false to leave the parameters on the stack
  * @return IteratorAggregate|boolean adodb result set or false if the query failed.
  */
-function db_query( $p_query, array $p_arr_parms = null, $p_limit = -1, $p_offset = -1 ) {
+function db_query( $p_query, array $p_arr_parms = null, $p_limit = -1, $p_offset = -1, $p_pop_param = true ) {
```

Comment:
> On a side note, shouldn't we also pop() when the query fails? (for the sake of correctness)

Made some debugging, and i'm positive it should also pop() on query fail, although the problem is quite subtle here.

First, note that an sql error is not a script halt. We can explicitly supress errors in api calls, e.g "@config_get(...)", or manage error recovery, exception, etc.

So, when the query fails:
1) ADOdb internal param count is resetted to 1 (pgsql)
2) $g_db_param param count is not cleared, so it will keep counting on next db_param()

**This happens in disconnected code**, not just in the nested query situation, as $g_db_param param is persistent and global

How can this be a failure point? i'm thinking Oracle, which uses named parameters, which mantis api names after its param count.
Probably what wil happen, on next db_param calls:
mysql: ?, ?, ? .. (ok)
pgsql: $1, $2, $3 ... (the number comes from ADOdb)
oracle: :5, :6, :7  (fail!)